### ISSUE TITLE
Feature request - OSC support

### ISSUE THREAD

**Issue by mrlimbic** [lhc70000/iina#1319]
It would be great to be able to remote control the player from another app via OSC (Open sound control) protocol.

Basic transport (play/stop) and locate frame. So be able to go to a specific timecode when stopped.

This would make it useful as an external player for another app, or over a network controlled from many apps that already support outputting OSC commands.

**Comment by krlklm** [comment]
Does anyone know if there is another movie player out there, where i can control the video file, the windows size and position and the playback position remotely? Movie player that come to mind: QuickTime Player, mpv.io, mplayer2, VLC...

**Comment by mrlimbic** [comment]
As well as OSC (Open Sound Control) protocol, MTC/MMC sync is also really useful. Most DAWs support sending MTC (midi timecode) for playback sync and MMC (midi machine code) for scrub/locate while editing (video is not playing back at that time but moving playhead in the DAW should still locate the correct frame).

@krlklm There are a few players that support remote playback sync via MTC/MMC but not the window size/position that I know of. They all assume you've already got the window how you want it but just need to control playback.

**Comment by krlklm** [comment]
It would be sufficient to define the window position and size before opening the window. After that i would like to control the playback position. I think i got it working with the mpv-player (mpv.io).

Now my client asks for rounded corners, similar to the quicktime player, but it doesn't seem to be possible with the mpv-player. IINA, which has rounded corners, on the other hand does not seem to be controllable via the terminal.

Is there any other controllable media player with rounded corners that i could check out?

**Comment by lhc70000** [comment]
IINA supports all mpv's features except for some window related options. If you can get it work with mpv, it should also work for IINA.

If you need a CLI for IINA, check out `/Applications/IINA.app/Contents/MacOS/iina-cli`.

**Comment by krlklm** [comment]
Seems like this is brand new. Great!

I found this on the release notes for 0.0.15:

> Command Line Tool
> 
> Finally, the command line tool for IINA is here. To enable it, simply add a symlink to/Applications/IINA.app/Contents/MacOS/iina-cli in your PATH:
> 
> ln -s /Applications/IINA.app/Contents/MacOS/iina-cli ~/bin/iina

I didn't find the folder ~/bin/. Could it be that the recommended location for the symlink should be /usr/local/bin/ ? I'm on 10.12.6 Sierra.

**Comment by lhc70000** [comment]
@krlklm You can link it to everywhere, as long as it's in your `PATH`.

**Comment by saagarjha** [comment]
@lhc70000 `~/bin` doesn't exist for most users, nor is on `PATH`. You may want to suggest installing it in `/usr/local/bin`.

**Comment by tolgraven** [comment]
Note: iina-cli (afaics) only allows launch, not communicating with a running instance.
For that, you can still use iina with mpv.conf and `input-ipc-server=/tmp/mpv-socket`, then `socat` your commands. More info in mpv docs.

So stuff like OSC, MTC/MMC, what have you, is probably best implemented by scripts translating between the various formats. Build a json, parse a json.
Example, my basic set up for easy applescript/BTT touchbar interaction, in `fish`.

```fish
function mpv-cmd --description 'control mpv/iina over socket' --argument cmd val type
  test -z "$type"; or test "$type" = "set"; and set type "set_property"
  not test -z "$val";  and set val ", $val"
  echo "{ \"command\": [\"$type\", \"$cmd\"$val]}" | socat - /tmp/mpv-socket
end

function mpv-info --description 'get info from mpv/iina over socket, via mpv-cmd' --argument cmd
  mpv-cmd $cmd '' 'get_property' | jq '.data'
end
```
<img width="471" alt="screen shot 2018-06-20 at 00 10 23" src="https://user-images.githubusercontent.com/9143357/41627762-e60343f8-7421-11e8-8af6-cc0f311fc89c.png">

**Comment by mrlimbic** [comment]
@talgraven Very interesting. How accurate is that method? The trouble with syncing players/apps for editing purposes is that unless it is frame accurate it doesn't really work for editing. Imagine trying to record dialogue in one app to an image in another app. The timing needs to be spot on for the recording to be useful.

**Comment by tolgraven** [comment]
Exact seek is available, yes.
https://mpv.io/manual/master/#list-of-input-commands

**Comment by mrlimbic** [comment]
OK. Very interesting. Needs a bit of testing. The only problem I can see of the top of my head is that seek position needs to be exact rational numbers, not floating point seconds. Frame duration in NTSC is 1001/30000 or 1001/24000 which doesn't translate well to floating point. FP often leads to rounding errors when stopping on frames. Really you to seek want by exact frame number (for a particular timebase), not time itself.

**Comment by mrlimbic** [comment]
This is the basic idea behind timecode. You are seeking actual frames, not a particular time in seconds. i.e. I know at 23.98fps (really 23.976023976023976) exactly what frame is meant by 01:05:03:02 with that particular timecode in 1001/24000 timebase. So one way of sorting this is to allow selecting seek positions by a specified timecode and timebase, instead of time only.

**Comment by tolgraven** [comment]
Right, it's obviously not tailored for a pro workflow. But between exact second seek, `estimated-frame-number` / `estimated-frame-count`, and `frame-step` / `frame-back-step`, I reckon a possibly passable solution could be found.

**Comment by mrlimbic** [comment]
Yeah. But it's so close to a perfect solution if you could pass the parameters as timecode and timebase to the app instead of seconds. The app obviously needs to implement ecoders/decoders for various standard frame rates but there are probably plenty of libraries that have those calculations already. If not, I'm happy to donate some code to incorporate. Here's a snippet which I actually got the maths of someone else.

	```
public long[] toHMSF(long frames) {
		if (ntsc) {
			if (drop) {
				if (timebase == 30) {
					// Formula from http://andrewduncan.net/timecodes/
					long D = frames / 17982;
					long M = frames % 17982;
					frames += 18*D + 2*((M - 2) / 1798);
//				    (If -2 div 1798 doesn't return 0, you'll have to special-case M = 0 or 1.)
					long f = frames % 30;
					long s = (frames / 30) % 60;
					long m =  ((frames / 30) / 60) % 60;
					long h   = (((frames / 30) / 60) / 60) % 24;
					return new long[] { h, m, s, f };
				} else {
					throw new UnsupportedOperationException("Drop frame not supported for timebase " + timebase);
				}
			} 
		}

		// TODO What about NTSC? Frame duration???
		Rational d = ntsc ? Rational.valueOf(1, timebase) : getFrameDuration(); //
		Rational t = d.times(frames);
		
		// Simple non drop for all other timebases
		long h = t.divide(Rational.valueOf(3600, 1)).longValue() % 24;
		long m = t.divide(Rational.valueOf(60, 1)).longValue() % 60;
		long s = t.longValue() % 60;
		long f = frames % timebase;
//		long h = frames / 3600 / timebase % 24;
//		long m = frames / 60 / timebase % 60;
//		long s = frames / timebase % 60;
//		long f = frames % timebase;	
		
		return new long[] { h, m, s, f };
	}

```

**Comment by tolgraven** [comment]
Right, I'm saying for frame-based seeking first at least try seek second, get `estimated-frame-number`, and do some frame-steps to get to where you really want. Maybe it'll be too off to be useful, maybe not. Only one way to find out - script it and try! :)

I'd still reckon this is pretty niche for a media-consumption target player, given pros generally use, well, pro stuff.

**Comment by mrlimbic** [comment]
Sure. The only difference between pro vs not pro for a video player is actually just a pretty tiny bit of math. Maybe a 100 lines at max to accept frame accurate seeks. There are some good free (but not open source ones) such as streamers (Mac OS only and is quite fussy about footage format). I am just suggesting that it would be nice if some open source video apps added a few lines to enable frame accurate seeking too.

**Comment by tolgraven** [comment]
Agreed! But again, for that to be possible you'd need to go to the mpv repo, not iina.

**Comment by Henrietta1989** [comment]
Is there any way to get current time of the movie in seconds?

And full path to video?

Macintosh HD/Videos/somevideo.mpg

**Comment by saagarjha** [comment]
I not sure this is completely related to this issue. However, to answer your question: if you’re just looking to see the file’s path, you can secondary click on the document proxy icon in the titlebar.

**Comment by Henrietta1989** [comment]
Thanks for answer, but i have AppleScript which needs path and current time.
